README.md
# Pushy messaging for Flutter

A Flutter plugin to use push notification delivery service from [PUSHY](https://pushy.me/).

With this plugin, your Flutter app can receive and process push notifications on Android and iOS.
Read [Pushy SDK documentation](https://pushy.me/docs) to get good understanding about how it works.

*Note*: This plugin is still under development, and some APIs might not be available yet.
[FEEDBACK](https://github.com/holmusk/flutter_pushy/issues) and [PULL REQUEST](https://github.com/holmusk/flutter_pushy/pulls) are most welcome!

## Usage
To use this plugin, add `flutter_pushy` as a [dependency in your pubspec.yaml file](https://flutter.io/platform-plugins/).

```
dependencies:
    flutter_pushy: ^1.0.0
```

## Getting Started
Check out the `example` directory for a sample app using Pushy.
Sign up on [Pushy website](https://pushy.me), to access dashboard page and create an app. 

## Android Integration

To integrate your plugin intot the Android part of your app, follow these steps:
1. Using Pushy console [create Android app](https://pushy.me/docs/android/create-app) in dashboard.
2. (Optional, but recommended) If want to be notified in your app (via onResume) when user clicks on a notification in the system tray, include the following code to your `AndroidManifest.xml`, inside activity tag.
```
<application>
    <activity ...>
        ...
        <intent-filter>
            <action android:name="PUSHY_NOTIFICATION_CLICK" />
            <category android:name="android.intent.category.DEFAULT" />
        </intent-filter>
    </activity>
</application>
```

## iOS Integration

To integrate your plugin into the iOS part of your app, follow these steps:
1. Using Pushy console [create Android app](https://pushy.me/docs/ios/create-app) in dashboard.
2. [Enable push notification capability](https://pushy.me/docs/ios/enable-capability) in Xcode project.
3. Setup [APNS authentication](https://pushy.me/docs/ios/setup-apns-auth) and export it's authentication key.

## Dart/Flutter Integration

From your Dart code, you need to import the plugin and instatiate it:
```
import 'package:flutter_pushy/flutter_pushy.dart';
final FlutterPushy _pushy = FlutterPushy();
```
> On Android this will bring up write external storage permissions for user to confirm. This is mandatory to prevent multiple token issued for same device for each installation.

### Register your device to Pushy server. 
> `_pushy.registerDevice()`. 

IOS - This will bring up a permissions dialog for user to confirm on. It's no-op on Android (*if write external storage permission is granted). Pushy will issue a token for each device once device registered, which will be handled in next step.

### Configure pushy plugin.
> `_pushy.configure(...)`
1. **onMessage**
triggered when on notification received when app is in foreground.
In iOS you must handle in-app notification manually. (ex. use LocalNotification or Dialog widget)
2. **onResume**
    Triggered on user tap on push notification when app is in background.
3. **onToken**
    Triggered if device registered successfully to Pushy server.
4. **onRegisterFail**
    Triggered if device registration failed, return `500` (network error) or `401` (write external storage permissions is denied - Android only).
    
```
_pushy.configure(
      onMessage: (data) {
      print('DATA ON MESSAGE: $data');
      if (LocalPlatform().isIOS) {
        ...
      }
      ...
    }, onResume: (data) {
      print('DATA ON RESUME: $data');
      ...
    },
     onRegisterFail: (err) {
     print('failed to register device with code :$err')
       ...
     }
    , onToken: (token) {
      print('found new Token: $token');
      ...
    });
```
## Sending Test Messages

Refer to Pushy [dashboard](https://dashboard.pushy.me/apps/<APP_ID>/send) or startup guide on [Android](https://pushy.me/docs/android/send-test-notification) or [iOS](https://pushy.me/docs/ios/send-test-notification) to send test notification.
> Make sure to set **Content-Available** as true for iOS




CHANGELOG.md
## 1.0.0

* Intial Flutter Pushy Pluggin.
* Request for write external storage permission on Android.
* Register device to pushy server and failure handling.
* Handle and trigger a callback when push notification received.
* Example project how to use this plugin.


test/flutter_pushy_test.dart
// import 'package:test/test.dart';
// import 'package:mockito/mockito.dart';
// import 'package:platform/platform.dart';
// import 'package:flutter/services.dart';
// import 'package:flutter_pushy/flutter_pushy.dart';

// void main() {
//   MockMethodChannel mockChannel;
//   FlutterPushy pushy;

//   setUp(() {
//     print('SETUP');
//     mockChannel = MockMethodChannel();
//     pushy       = FlutterPushy.private(mockChannel, FakePlatform(operatingSystem: 'ios'));
//   });

//   test('reguestWriteExternalStoragePermission_IOS', () {
//     pushy.requestWriteExtStoragePermission();
//     verifyNever(mockChannel.setMethodCallHandler(any));
//     verifyNever(mockChannel.invokeMethod('requestWriteExtStoragePermission'));
//   });

//   test('fetchWriteExternalStoragePermission_IOS', () async {
//     final status = await pushy.writeExtStoragePermission;
//     verifyNever(mockChannel.setMethodCallHandler(any));
//     verifyNever(mockChannel.invokeMethod('requestWriteExtStoragePermission'));
//     assert(status, equals(true));
//   });

//   test('reguestWriteExternalStoragePermission_Android', () {
//     pushy = FlutterPushy.private(mockChannel, FakePlatform(operatingSystem: 'android'));
//     pushy.requestWriteExtStoragePermission();
//     verifyNever(mockChannel.setMethodCallHandler(any));
//     verify(mockChannel.invokeMethod('requestWriteExtStoragePermission'));
//   });

//   test('fetchWriteExternalStoragePermission_Android', () async {
//     pushy = FlutterPushy.private(mockChannel, FakePlatform(operatingSystem: 'android'));
//     final bool denied = false;
//     verifyNever(mockChannel.invokeMethod('fetchWriteExtStoragePermission', denied));
//     mockChannel.invokeMethod('fetchWriteExtStoragePermission', denied);
//     verify(mockChannel.invokeMethod('fetchWriteExtStoragePermission', denied));
//     final bool granted = true;
//     verifyNever(mockChannel.invokeMethod('fetchWriteExtStoragePermission', granted));
//     mockChannel.invokeMethod('fetchWriteExtStoragePermission', granted);
//     verify(mockChannel.invokeMethod('fetchWriteExtStoragePermission', granted));
//   });



//   test('registerDevice', () {
//     pushy.registerDevice();
//     verify(mockChannel.invokeMethod('registerDevice'));
//   });

//   test('configure', () {
//     pushy.configure();
//     verify(mockChannel.setMethodCallHandler(any));
//     verify(mockChannel.invokeMethod('configure'));
//   });
// }

// class MockMethodChannel extends Mock implements MethodChannel {}


example/test/widget_test.dart
// This is a basic Flutter widget test.
//
// To perform an interaction with a widget in your test, use the WidgetTester
// utility that Flutter provides. For example, you can send tap and scroll
// gestures. You can also use WidgetTester to find child widgets in the widget
// tree, read text, and verify that the values of widget properties are correct.

import 'package:flutter/material.dart';
import 'package:flutter_test/flutter_test.dart';

import 'package:flutter_pushy_example/main.dart';

void main() {
  testWidgets('Verify Platform version', (WidgetTester tester) async {
    // Build our app and trigger a frame.
    await tester.pumpWidget(MyApp());

    // Verify that platform version is retrieved.
    expect(
      find.byWidgetPredicate(
        (Widget widget) => widget is Text &&
                           widget.data.startsWith('Running on:'),
      ),
      findsOneWidget,
    );
  });
}


example/lib/main.dart
import 'package:flutter/material.dart';
// import 'package:flutter_local_notifications/flutter_local_notifications.dart';
import 'package:flutter_pushy/flutter_pushy.dart';
import 'package:platform/platform.dart';


void main() => runApp(MyApp());

class MyApp extends StatefulWidget {
  @override
  _MyAppState createState() => _MyAppState();
}

class _MyAppState extends State<MyApp> {
  String _data;
  String _token = '';
  final _pushy = FlutterPushy();
  bool _isLoading = false;

  // FlutterLocalNotificationsPlugin _localPlugin = FlutterLocalNotificationsPlugin();

  @override
  void initState() {
    super.initState();
    _fetchToken();
    _configurePushy();
    // _configureLocalNotification();
  }

  void _configurePushy() {
    _pushy.configure(
      onMessage: (data) {
      print('DATA ON MESSAGE: $data');
      setState(() => _data = data.toString());
      // Handle onMessage for iOS manually
      if (LocalPlatform().isIOS) {
        // final body = data['message'] ?? 'Notification';
        // showNotification(body, data.toString());
      }
    }, onResume: (data) {
      print('DATA ON RESUME: $data');
      setState(() => _data = data.toString());
    },
     onRegisterFail: (err) {
       print('FAILED TO REGISTER DEVICE WITH CODE: $err');
     }
    , onToken: (token) {
      print('found new Token: $token');
      setState(() => _token = token);
    });
  }

  void _fetchToken() async {
    final token = await FlutterPushy().getToken();
    print('Token: $token');
    setState(() => _token = token);
  }

  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      home: Scaffold(
        appBar: AppBar(
          title: const Text('Pushy Test'),
        ),
        body: LocalPlatform().isIOS
            ? _iOScontent(context)
            : _iAndroidDontent(context),
      ),
    );
  }

  Widget _iOScontent(BuildContext context) {
    return _isLoading
        ? Center(child: CircularProgressIndicator())
        : Padding(
            padding: const EdgeInsets.all(24.0),
            child: Column(
              mainAxisSize: MainAxisSize.max,
              children: <Widget>[
                RaisedButton(
                  child: Text('Register Device'),
                  onPressed: () => _pushy.registerDevice(),
                ),
                SizedBox(height: 24.0),
                _tokenWidget(context),
                SizedBox(height: 24.0),
                (_data == null)
                    ? Container()
                    : Text(
                        'Last Data Sent:\n${_data.toString()}',
                        textAlign: TextAlign.center,
                      ),
              ],
            ),
          );
  }

  Widget _iAndroidDontent(BuildContext context) {
    return _isLoading
        ? Center(child: CircularProgressIndicator())
        : Padding(
            padding: const EdgeInsets.all(24.0),
            child: Column(
              mainAxisSize: MainAxisSize.max,
              children: <Widget>[
                RaisedButton(
                  child: Text('Request write permission'),
                  onPressed: () {
                    _pushy.requestWriteExtStoragePermission();
                  },
                ),
                RaisedButton(
                  child: Text('Get write permission'),
                  onPressed: () async {
                    final granted = await _pushy.writeExtStoragePermission;
                    print(
                        'Write external storage permission: ${granted ? 'yes' : 'no'}');
                  },
                ),
                RaisedButton(
                  child: Text('Register Device'),
                  onPressed: () => _pushy.registerDevice(),
                ),
                SizedBox(height: 24.0),
                _tokenWidget(context),
                SizedBox(height: 24.0),
                (_data == null)
                    ? Container()
                    : Text(
                        'Last Data Sent:\n${_data.toString()}',
                        textAlign: TextAlign.center,
                      ),
              ],
            ),
          );
  }

  Widget _tokenWidget(BuildContext context) {
    return (_token == null)
        ? Container()
        : Text(
            'Device Token:\n$_token',
            textAlign: TextAlign.center,
          );
  }


  /* 
  IOS ONLY!!!
  Example of how to handle onMessage with Local Notification
  
  void _configureLocalNotification() {
    var initializationSettingsAndroid =
    AndroidInitializationSettings('app_icon');
    
    var initializationSettingsIOS = IOSInitializationSettings(
    onDidReceiveLocalNotification: _onDidReceiveLocalNotification);
    
    var initializationSettings = InitializationSettings(
    initializationSettingsAndroid, initializationSettingsIOS);
    
    _localPlugin.initialize(initializationSettings,
    onSelectNotification: _onSelectNotification);
  }

  Future<void> onDidReceiveLocalNotification(
    int id,
   String title, 
   String body, 
   String payload) async {
     print('RECEIVE LOCAL NOTIF PAYLOAD : $payload');  
  }
  Future<void> _onDidReceiveLocalNotification(
    int id,
   String title, 
   String body, 
   String payload) async {
     print('RECEIVE LOCAL NOTIF PAYLOAD : $payload');  
  }

  Future<void> _onSelectNotification(String payload) async {
    print('HANDLE LOCAL NOTIFICATION SELECTION $payload');
  }

  Future<void> showNotification(String body, String payload) async {
    print('SHOW LOCAL NOTIFICATION $body $payload');
    var androidPlatformChannelSpecifics = AndroidNotificationDetails(
        'com.holmusk.glycoleap.inAppLocal',
        'inAppLocal',
        'In App Local notification',
        importance: Importance.Max,
        priority: Priority.High);
    var iOSPlatformChannelSpecifics = IOSNotificationDetails();
    var platformChannelSpecifics = NotificationDetails(
        androidPlatformChannelSpecifics, iOSPlatformChannelSpecifics);
    await _localPlugin.show(0, 'GLYCO', body, platformChannelSpecifics,
        payload: payload);
  }
  */
}


lib/flutter_pushy.dart
import 'dart:async';
import 'package:flutter/material.dart';
import 'package:flutter/services.dart';
import 'package:meta/meta.dart';
import 'package:platform/platform.dart';


/// Event handler types
typedef Future<dynamic> MessageHandler(Map<String, dynamic> message);
typedef Future<String>  TokenHandler(String token);
typedef Future<int>     ErrorHandler(int errorCode );

/// Implementation of pushy messaging API for flutter
///
/// Your app should call [registerDevice] to get a token
/// then register handles dor incoming messages with [configure]
///
class FlutterPushy {
  factory FlutterPushy() => _instance;

  @visibleForTesting
  FlutterPushy.private(MethodChannel channel, Platform platform,)
      : _channel = channel,
        _platform = platform;

  static final FlutterPushy _instance = 
  FlutterPushy.private(const MethodChannel('com.holmusk.flutter_pushy'), 
                       const LocalPlatform());

  final MethodChannel _channel;
  final Platform      _platform;
  MessageHandler      _onMessage ;
  MessageHandler      _onResume;
  ErrorHandler        _onRegisterFail;
  TokenHandler        _onToken;

  
  /// Request write external storage permissions <Android only>
  Future<void> requestWriteExtStoragePermission() async {
    // This is mandatory to prevent multiple token issued for same device.
    if (_platform.isAndroid) {
      return _channel.invokeMethod('requestWriteExtStoragePermission');
    }
  }
  
  /// Get status of write external storage permission <Android only>
  Future<bool> get writeExtStoragePermission async {
    // if not android return true
    if (!_platform.isAndroid) { return true; }
    return _channel.invokeMethod('fetchWriteExtStoragePermission');
  }
    
  /// Sets up [Handlers] for incoming events.
  void configure({
    MessageHandler onMessage, 
    MessageHandler onResume, 
    TokenHandler onToken, 
    ErrorHandler onRegisterFail}) {
      _onToken        = onToken;
      _onMessage      = onMessage;
      _onResume       = onResume;
      _onRegisterFail = onRegisterFail;
      _channel.setMethodCallHandler(_handleMethod);
      _channel.invokeMethod('configure');
  }

  /// Register device to pushy server
  Future<Null> registerDevice() async => await _channel.invokeMethod('registerDevice');

  /// Returns locally stored [token] String
  Future<String> getToken() async => await _channel.invokeMethod('getToken');
  
  /// Handle callback from plugin itself
  Future<dynamic> _handleMethod(MethodCall call) async {
    switch (call.method) {
      case 'onMessage':
        return _onMessage(call.arguments.cast<String, dynamic>());
      case 'onResume':
        return _onResume(call.arguments.cast<String, dynamic>());
      case 'onRegisterFail':
        return _onRegisterFail(call.arguments);
      case 'onToken':
        return _onToken(call.arguments);
      default:
        throw UnsupportedError('Unrecognized JSON message');
    }
  }
}

ios/Classes/SwiftFlutterPushyPlugin.swift
import Flutter
import UIKit
import Pushy

public class SwiftFlutterPushyPlugin: NSObject, FlutterPlugin {
  
  static let PUSHY_TOKEN_KEY = "pushyToken"
  static let FLUTTER_CHANNEL = "com.holmusk.flutter_pushy"
  
  var _pushy                  : Pushy
  var _channel                : FlutterMethodChannel
  var _resumingFromBackground : Bool
  var _resumingData           : Dictionary<AnyHashable, Any>?
  
  init(channel: FlutterMethodChannel) {
    _channel  = channel
    _pushy    = Pushy(UIApplication.shared)
    _resumingFromBackground = false;
  }
  
  /// =======================================================================================
  /// Pushy methods
  /// =======================================================================================
  public static func register(with registrar: FlutterPluginRegistrar) {
    let channel = FlutterMethodChannel(name: FLUTTER_CHANNEL,
                                       binaryMessenger: registrar.messenger())
    let instance = SwiftFlutterPushyPlugin(channel: channel)
    registrar.addMethodCallDelegate(instance, channel: channel)
    registrar.addApplicationDelegate(instance)
  }
  
  public func handle(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
    let method = call.method
    switch method {
    case "registerDevice":
      registerPushy()
      result(nil)
    case "configure":
      result(nil)
    case "getToken":
      result(UserDefaults.standard.string(forKey: SwiftFlutterPushyPlugin.PUSHY_TOKEN_KEY))
    default:
      result(FlutterMethodNotImplemented)
    }
  }
  
  /// =======================================================================================
  /// Private methods
  /// =======================================================================================
  private func registerPushy() {
    _pushy.register({ [unowned self](error, deviceToken) in
      if error == nil {
        UserDefaults.standard.set(deviceToken,
                                  forKey: SwiftFlutterPushyPlugin.PUSHY_TOKEN_KEY)
        self._channel.invokeMethod("onToken", arguments: deviceToken)
      } else {
        self._channel.invokeMethod("onRegisterFail", arguments: 500)
      }
    })
  }
  
  private func setNotificationHandler() {
    _pushy.setNotificationHandler({ [unowned self](data, completionHandler) in
      self._resumingFromBackground ?
        // set data to invoke onResume when app became active
        self._resumingData = data :
        // invoke onMessage directly when app is active
        self._channel.invokeMethod("onMessage", arguments: data)
      completionHandler(UIBackgroundFetchResult.newData)
    })
  }
  
  /// =======================================================================================
  /// Application delegates
  /// =======================================================================================
  public func application(_ application: UIApplication,
                          didFinishLaunchingWithOptions launchOptions: [AnyHashable : Any] = [:]) -> Bool {
    setNotificationHandler()
    return true
  }
  
  public func applicationDidBecomeActive(_ application: UIApplication) {
    _resumingFromBackground = false
    // Reset badge number once app goes to foreground
    application.applicationIconBadgeNumber = 1;
    application.applicationIconBadgeNumber = 0;
    // If _resumingData is not nil then invoke onResume
    if let data = _resumingData {
      _channel.invokeMethod("onResume", arguments: data)
      _resumingData = nil
    }
  }
  
  public func applicationDidEnterBackground(_ application: UIApplication) {
    _resumingFromBackground = true
  }
}


android/src/main/java/com/holmusk/flutter_pushy/FlutterPushyPlugin.java
package com.holmusk.flutter_pushy;

import io.flutter.plugin.common.FlutterException;
import me.pushy.sdk.Pushy;

import java.util.HashMap;
import java.util.Map;
import java.util.Set;
import java.lang.ref.WeakReference;

import io.flutter.plugin.common.FlutterException;
import io.flutter.plugin.common.MethodCall;
import io.flutter.plugin.common.MethodChannel;
import io.flutter.plugin.common.MethodChannel.MethodCallHandler;
import io.flutter.plugin.common.MethodChannel.Result;
import io.flutter.plugin.common.PluginRegistry.Registrar;
import io.flutter.plugin.common.PluginRegistry.NewIntentListener;

import android.os.AsyncTask;
import android.os.Bundle;

import android.util.Log;
import android.media.RingtoneManager;
import android.graphics.Color;
import android.preference.PreferenceManager;

import android.app.PendingIntent;
import android.app.NotificationManager;

import android.content.SharedPreferences;
import android.content.Context;
import android.content.pm.PackageManager;
import android.content.Intent;
import android.content.IntentFilter;
import android.content.BroadcastReceiver;

import android.support.v4.content.ContextCompat;
import android.support.v4.app.ActivityCompat;
import android.support.v4.app.NotificationCompat;
import android.support.v4.content.LocalBroadcastManager;


public class FlutterPushyPlugin
  extends BroadcastReceiver
  implements MethodCallHandler, NewIntentListener {

  private static final String CLICK_ACTION_VALUE = "PUSHY_NOTIFICATION_CLICK";
  private static final String FLUTTER_CHANNEL = "com.holmusk.flutter_pushy";

  private final MethodChannel channel;
  private final Registrar registrar;

  RegisterDeviceTask _runner;

  /// =====================================================================================
  /// Flutter plugin methods
  /// =====================================================================================
  public static void registerWith(Registrar registrar) {
    final MethodChannel channel = 
      new MethodChannel(registrar.messenger(), FLUTTER_CHANNEL);
    final FlutterPushyPlugin plugin = 
      new FlutterPushyPlugin(registrar, channel);
    channel.setMethodCallHandler(plugin);
    plugin.requestWriteExtStoragePermission();
    Pushy.listen(registrar.activeContext());
    registrar.addNewIntentListener(plugin);
  }

  private FlutterPushyPlugin(Registrar registrar, MethodChannel channel) {
    this.registrar  = registrar;
    this.channel    = channel;
    IntentFilter intentFilter = new IntentFilter();
    intentFilter.addAction(PushReceiver.ACTION_REMOTE_MESSAGE);
    LocalBroadcastManager manager = LocalBroadcastManager.getInstance(registrar.activeContext());
    manager.registerReceiver(this, intentFilter);
  }

  @Override
  public void onMethodCall(final MethodCall call, final Result result) {
    switch (call.method) {
      case "registerDevice":
        registerDevice();
        result.success(null);
        break;
      case "getToken":
        String deviceToken = getDeviceToken();
        result.success(deviceToken);
        break;
      case "configure":
        result.success(null);
        break;
      case "requestWriteExtStoragePermission":
        this.requestWriteExtStoragePermission();
        result.success(null);
        break;
      case "fetchWriteExtStoragePermission":
        result.success(isWriteExtStorageGranted());
        break;
      default:
        result.notImplemented();
        break;
    }
  }

  /// =====================================================================================
  /// Register device methods
  /// =====================================================================================
  private void registerDevice() {
    // Register device only if write external storage permission is granted
    // to prevent multiple token registered from same device
    // Pushy billed per device based!
    if (isWriteExtStorageGranted()) {
      if (_runner != null) {
        _runner.cancel(true);
      }
      _runner = new RegisterDeviceTask(this);
      _runner.execute();
    } else {
      channel.invokeMethod("onRegisterFail", 401);
    }
  }

  /// =====================================================================================
  /// Write external storage permissions methods
  /// =====================================================================================
  private boolean isWriteExtStorageGranted() {
    return ContextCompat.checkSelfPermission(
            this.registrar.activeContext(), android.Manifest.permission.WRITE_EXTERNAL_STORAGE)
            == PackageManager.PERMISSION_GRANTED;
  }

  /// Request Write External Storage Permission
  private void requestWriteExtStoragePermission() {
    if (!isWriteExtStorageGranted()) {
      ActivityCompat.requestPermissions(this.registrar.activity(),
              new String[]{android.Manifest.permission.READ_EXTERNAL_STORAGE,
                      android.Manifest.permission.WRITE_EXTERNAL_STORAGE}, 0);
    }
  }

  /// =====================================================================================
  ///  Intent receiving methods
  /// =====================================================================================
  @Override
  public void onReceive(Context context, Intent intent) {

    if (intent.getAction() != null
            && !intent.getAction().equals(PushReceiver.ACTION_REMOTE_MESSAGE)) {
      Log.d("Warning", "Received unknown action");
      return;
    }

    String title = intent.getStringExtra("title") != null ?
            intent.getStringExtra("title") :
            "Notification";
    String message = intent.getStringExtra("message") != null ?
            intent.getStringExtra("message") :
            "You receive a notification";

    final Bundle bundle = intent.getExtras();
    if (bundle == null) {
      Log.d("Warning", "Received message without payload");
      return;
    }
    channel.invokeMethod("onMessage", bundleToMap(bundle));

    int NOTIFICATION_ID = 234;
    String CHANNEL_ID = "flutter_pushy";

    Intent nIntent = new Intent(CLICK_ACTION_VALUE);
    nIntent.putExtras(bundle);
    PendingIntent pIntent = PendingIntent.getActivity(context,
            0,
            nIntent,
            PendingIntent.FLAG_UPDATE_CURRENT);

    NotificationCompat.Builder builder =
    new NotificationCompat.Builder(context, CHANNEL_ID)
    .setAutoCancel(true)
    .setSmallIcon(android.R.drawable.ic_dialog_info)
    .setLights(Color.RED, 1000, 1000)
    .setVibrate(new long[]{0, 400, 250, 400})
    .setSound(RingtoneManager.getDefaultUri(RingtoneManager.TYPE_NOTIFICATION))
    .setContentTitle(title)
    .setContentText(message)
    .setContentIntent(pIntent);

    Pushy.setNotificationChannel(builder, context);

    NotificationManager manager =
            (NotificationManager) context.getSystemService(Context.NOTIFICATION_SERVICE);
    manager.notify(NOTIFICATION_ID, builder.build());
  }

  @Override
    public boolean onNewIntent(Intent intent) {
      Log.d(intent.getAction(), "On intent received from local notification");

      if (intent.getAction() != null && !intent.getAction().equals(CLICK_ACTION_VALUE)) {
        Log.d("Warning", "Received unknown action");
        return false;
      }

      final Bundle bundle = intent.getExtras();
      if (bundle == null) {
        Log.d("Warning", "Received message without payload");
        return false;
      }
      channel.invokeMethod("onResume", bundleToMap(bundle));
      return true;
    }

  /// =====================================================================================
  ///  Methods to store and fetch token from local preferences
  /// =====================================================================================
  private String getDeviceToken() {
    return getSharedPreferences().getString("deviceToken", null);
  }

  private void saveDeviceToken(String deviceToken) {
    getSharedPreferences().edit().putString("deviceToken", deviceToken).apply();
  }

  private SharedPreferences getSharedPreferences() {
    return PreferenceManager.getDefaultSharedPreferences(this.registrar.activeContext());
  }

  /// =====================================================================================
  /// Register device asynchronous runner
  /// =====================================================================================
  private static class RegisterDeviceTask extends AsyncTask<Void, Void, String> {

    private WeakReference<FlutterPushyPlugin> activityReference;

    RegisterDeviceTask(FlutterPushyPlugin context) {
      activityReference = new WeakReference<>(context);
    }


    @Override
    protected String doInBackground(Void... params) {
      FlutterPushyPlugin activity = activityReference.get();
      if (activity == null) {
        return null;
      }
      try {
        return Pushy.register(activity.registrar.activeContext());
      } catch (Exception e) {
        Log.d("RegisterDeviceTask", e.getMessage());
        return null;
      }
    }

    @Override
    protected void onPostExecute(String result) {
      FlutterPushyPlugin activity = activityReference.get();
      if (activity != null) {
        if (result != null) {
          activity.saveDeviceToken(result);
          activity.channel.invokeMethod("onToken", result);
        } else {
          activity.channel.invokeMethod("onRegisterFail", 500);
        }
      }
    }
  }

  /// =====================================================================================
  /// Bundle to map methods
  /// =====================================================================================
  public static Map<String, Object> bundleToMap(Bundle extras) {
    Map<String, Object> map = new HashMap<>();
    Set<String> ks = extras.keySet();
    for (String key: ks) {
      final Object value = extras.get(key);
      Log.d("DEBUG", "");
      if (value != null) {
        map.put(key, value);
      }
    }
    return map;
  }
}
